homeassistant/components/control4/director_utils.py

```diff
@@ -0,0 +1,65 @@
+"""Provides data updates from the Control4 controller for platforms."""
+import logging
+
+from pyControl4.account import C4Account
+from pyControl4.director import C4Director
+from pyControl4.error_handling import BadToken
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_TOKEN, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import aiohttp_client
+
+from .const import (
+    CONF_ACCOUNT,
+    CONF_CONTROLLER_UNIQUE_ID,
+    CONF_DIRECTOR,
+    CONF_DIRECTOR_TOKEN_EXPIRATION,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def director_update_data(
+    hass: HomeAssistant, entry: ConfigEntry, var: str
+) -> dict:
+    """Retrieve data from the Control4 director for update_coordinator."""
+    # possibly implement usage of director_token_expiration to start
+    # token refresh without waiting for error to occur
+    try:
+        director = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR]
+        data = await director.getAllItemVariableValue(var)
+    except BadToken:
+        _LOGGER.info("Updating Control4 director token")
+        await refresh_tokens(hass, entry)
+        director = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR]
+        data = await director.getAllItemVariableValue(var)
+    return_dict = {}
+    for key in data:
+        return_dict[key["id"]] = key
+    return return_dict
```

Comment:
Consider a dict comprehension
```suggestion
    return {key["id"]: key for key in data}
```